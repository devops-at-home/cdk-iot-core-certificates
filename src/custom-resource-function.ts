// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for CustomResourceFunction
 */
export interface CustomResourceFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/custom-resource.
 */
export class CustomResourceFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: CustomResourceFunctionProps) {
    super(scope, id, {
      description: 'src/custom-resource.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs14.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/custom-resource.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}