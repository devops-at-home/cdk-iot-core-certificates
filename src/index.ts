import * as cdk from "@aws-cdk/core";
import * as iam from "@aws-cdk/aws-iam";
import * as lambda from "@aws-cdk/aws-lambda";
import { NodejsFunction } from "@aws-cdk/aws-lambda-nodejs";
import { Provider } from "@aws-cdk/custom-resources";
import { RetentionDays } from "@aws-cdk/aws-logs";
import * as ssm from "@aws-cdk/aws-ssm";
import { CfnCustomResource } from "@aws-cdk/aws-cloudformation";

export interface ThingWithCertProps {
  readonly deviceName: string;
  readonly saveToParamStore?: boolean;
  readonly paramPrefix?: string;
}

export class ThingWithCert extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: ThingWithCertProps) {
    super(scope, id);

    const { deviceName, saveToParamStore, paramPrefix } = props;

    const lambdaExecutionRole = new iam.Role(this, "LambdaExecutionRole", {
      assumedBy: new iam.CompositePrincipal(
        new iam.ServicePrincipal("lambda.amazonaws.com")
      ),
    });

    lambdaExecutionRole.addToPolicy(
      new iam.PolicyStatement({
        resources: ["arn:aws:logs:*:*:*"],
        actions: [
          "logs:CreateLogGroup",
          "logs:CreateLogStream",
          "logs:PutLogEvents",
        ],
      })
    );

    lambdaExecutionRole.addToPolicy(
      new iam.PolicyStatement({
        resources: ["*"],
        actions: ["iot:*"],
      })
    );

    const lambdaFunction = new NodejsFunction(this, "lambdaFunction", {
      entry: `${__dirname}/lambda/index.ts`,
      handler: "handler",
      runtime: lambda.Runtime.NODEJS_14_X,
      timeout: cdk.Duration.seconds(10),
      role: lambdaExecutionRole,
      logRetention: RetentionDays.ONE_DAY,
    });

    const lambdaProvider = new Provider(this, "lambdaProvider", {
      onEventHandler: lambdaFunction,
    });

    const lambdaCustomResource = new CfnCustomResource(
      this,
      "lambdaCustomResource",
      {
        serviceToken: lambdaProvider.serviceToken,
      }
    );

    lambdaCustomResource.addPropertyOverride("ThingName", deviceName);

    if (saveToParamStore) {
      new ssm.CfnParameter(this, "paramStoreCertPem", {
        type: "String",
        value: lambdaCustomResource.getAtt("certPem").toString(),
        name: `${paramPrefix}/${deviceName}/certPem`,
      });

      new ssm.CfnParameter(this, "paramStorePrivKey", {
        type: "String",
        value: lambdaCustomResource.getAtt("privKey").toString(),
        name: `${paramPrefix}/${deviceName}/privKey`,
      });
    }

    new cdk.CfnOutput(this, "thingArn", {
      value: lambdaCustomResource.getAtt("thingArn").toString(),
    });

    new cdk.CfnOutput(this, "certId", {
      value: lambdaCustomResource.getAtt("certId").toString(),
    });

    new cdk.CfnOutput(this, "certPem", {
      value: lambdaCustomResource.getAtt("certPem").toString(),
    });

    new cdk.CfnOutput(this, "privKey", {
      value: lambdaCustomResource.getAtt("privKey").toString(),
    });
  }
}
