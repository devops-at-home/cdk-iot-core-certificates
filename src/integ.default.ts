import * as cdk from '@aws-cdk/core';
import { ThingWithCert } from './';

export class IntegTesting {
  readonly stack: cdk.Stack[];
  constructor() {
    const app = new cdk.App();

    const env = {
      region:
        process.env.CDK_INTEG_REGION ||
        process.env.CDK_DEFAULT_REGION ||
        'us-east-1',
      account: process.env.CDK_INTEG_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new cdk.Stack(app, 'integ-stack', { env });

    new ThingWithCert(stack, 'ThingWithCert', {
      thingName: 'thingName',
      saveToParamStore: true,
    });

    this.stack = [stack];
  }
}

new IntegTesting();
