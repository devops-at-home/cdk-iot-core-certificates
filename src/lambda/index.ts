import * as lambda from "aws-lambda";
import { Iot } from "aws-sdk";
import { thingAdaptor } from "../lambda/adapters/thing";
import { iotAdaptor } from "../lambda/adapters/iot";

type Success = lambda.CloudFormationCustomResourceSuccessResponse;
type Failure = lambda.CloudFormationCustomResourceFailedResponse;

const thingHandler = thingAdaptor(iotAdaptor(new Iot()));

export const handler = async (
  event: lambda.CloudFormationCustomResourceEvent
): Promise<Success | Failure> => {
  console.log(`Received event: ${JSON.stringify(event)}`);
  try {
    const thingName = event.ResourceProperties["ThingName"];
    if (event.RequestType === "Create") {
      const { thingArn, certId, certPem, privKey } = await thingHandler.create(
        thingName
      );
      return {
        Status: "SUCCESS",
        PhysicalResourceId: thingArn,
        LogicalResourceId: event.LogicalResourceId,
        RequestId: event.RequestId,
        StackId: event.StackId,
        Data: {
          certPem: certPem,
          privKey: privKey,
          certId: certId,
        },
      };
    } else if (event.RequestType === "Delete") {
      await thingHandler.delete(thingName);
      return {
        Status: "SUCCESS",
        PhysicalResourceId: event.PhysicalResourceId,
        LogicalResourceId: event.LogicalResourceId,
        RequestId: event.RequestId,
        StackId: event.StackId,
      };
    } else if (event.RequestType === "Update") {
      console.log(`Updating thing: ${thingName}`);
      return {
        Status: "SUCCESS",
        PhysicalResourceId: event.PhysicalResourceId,
        LogicalResourceId: event.LogicalResourceId,
        RequestId: event.RequestId,
        StackId: event.StackId,
      };
    } else {
      throw new Error("Received invalid request type");
    }
  } catch (err) {
    return {
      Status: "FAILED",
      Reason: err.message,
      RequestId: event.RequestId,
      StackId: event.StackId,
      LogicalResourceId: event.LogicalResourceId!,
      // @ts-ignore
      PhysicalResourceId: event.PhysicalResourceId || event.LogicalResourceId,
    };
  }
};
