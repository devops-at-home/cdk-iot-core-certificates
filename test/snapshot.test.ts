import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ThingWithCert } from '../src';

describe('Snapshot test', () => {
  const stack = new Stack();

  new ThingWithCert(stack, 'ThingWithCert', {
    thingName: 'thingName',
    saveToParamStore: true,
    paramPrefix: 'paramPrefix',
  });

  const template = Template.fromStack(stack);

  test('Snapshot test', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
