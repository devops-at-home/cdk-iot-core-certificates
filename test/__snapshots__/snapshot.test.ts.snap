// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Snapshot test 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e5032a7ed03a7762b99560439c17f58d046a35d40ebf64ba70c63157df388563.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ThingWithCertLambdaExecutionRole25FAB8F4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ThingWithCertLambdaExecutionRoleDefaultPolicy9FF9FB4B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:*",
            },
            Object {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ThingWithCertLambdaExecutionRoleDefaultPolicy9FF9FB4B",
        "Roles": Array [
          Object {
            "Ref": "ThingWithCertLambdaExecutionRole25FAB8F4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ThingWithCertlambdaCustomResource28E12CE8": Object {
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "ThingWithCertlambdaProviderframeworkonEvent497EC884",
            "Arn",
          ],
        },
        "ThingName": "thingName",
      },
      "Type": "AWS::CloudFormation::CustomResource",
    },
    "ThingWithCertlambdaFunction9E90C8BA": Object {
      "DependsOn": Array [
        "ThingWithCertLambdaExecutionRoleDefaultPolicy9FF9FB4B",
        "ThingWithCertLambdaExecutionRole25FAB8F4",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a5630afc5926ce88b06077285993ada90c4cf71bd44f0c0b489a24bc743b37a8.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ThingWithCertLambdaExecutionRole25FAB8F4",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ThingWithCertlambdaFunctionLogRetention1BE9E451": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "ThingWithCertlambdaFunction9E90C8BA",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "ThingWithCertlambdaProviderframeworkonEvent497EC884": Object {
      "DependsOn": Array [
        "ThingWithCertlambdaProviderframeworkonEventServiceRoleDefaultPolicyC133ABFF",
        "ThingWithCertlambdaProviderframeworkonEventServiceRole95EF1588",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "eaeb69bc290b516fe3b049f89d6118b22249df682fbabf56af300cf345198574.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Default/ThingWithCert/lambdaProvider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "ThingWithCertlambdaFunction9E90C8BA",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ThingWithCertlambdaProviderframeworkonEventServiceRole95EF1588",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ThingWithCertlambdaProviderframeworkonEventServiceRole95EF1588": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ThingWithCertlambdaProviderframeworkonEventServiceRoleDefaultPolicyC133ABFF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "ThingWithCertlambdaFunction9E90C8BA",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "ThingWithCertlambdaFunction9E90C8BA",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ThingWithCertlambdaProviderframeworkonEventServiceRoleDefaultPolicyC133ABFF",
        "Roles": Array [
          Object {
            "Ref": "ThingWithCertlambdaProviderframeworkonEventServiceRole95EF1588",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ThingWithCertparamStoreCertPem15876E53": Object {
      "Properties": Object {
        "Name": "paramPrefix/thingName/certPem",
        "Type": "String",
        "Value": Object {
          "Fn::GetAtt": Array [
            "ThingWithCertlambdaCustomResource28E12CE8",
            "certPem",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ThingWithCertparamStorePrivKey2E5E90A6": Object {
      "Properties": Object {
        "Name": "paramPrefix/thingName/privKey",
        "Type": "String",
        "Value": Object {
          "Fn::GetAtt": Array [
            "ThingWithCertlambdaCustomResource28E12CE8",
            "privKey",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
